client_auto.go

```diff
@@ -0,0 +1,95 @@
+package hdfs
+
+import (
+	"encoding/xml"
+	"errors"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+)
+
+type Property struct {
+	Name  string `xml:"name"`
+	Value string `xml:"value"`
+}
+
+type Result struct {
+	Property []Property `xml:"property"`
+}
+
+type HadoopConf map[string]string
+
+var ErrUnresolvedNamenode = errors.New("Couldn't find a namenode address in any config.")
+
+// Get Hadoop Properties - try to open a conf file, marshal the results
+// into a Result object and return the Properties of that object.
+func LoadHadoopConfig(path string) (HadoopConf, error) {
+	result := Result{}
+	f, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	xmlErr := xml.Unmarshal(f, &result)
+	if xmlErr != nil {
+		return nil, xmlErr
+	}
+
+	hadoopConf := make(HadoopConf)
+	for _, prop := range result.Property {
+		hadoopConf[prop.Name] = prop.Value
+	}
+	return hadoopConf, nil
+}
+
+func (conf HadoopConf) Namenodes() []string {
+	var nns []string
+	for key, value := range conf {
+		if strings.Contains(key, "fs.defaultFS") {
+			nnUrl, _ := url.Parse(value)
+			nns = append(nns, nnUrl.Host)
+		}
+		if strings.HasPrefix(key, "dfs.namenode.rpc-address") {
+			nns = append(nns, value)
+		}
```

Comment:
This method should dedupe, I think. You can do that by writing to a `map[string]bool` first, and then pulling out the keys into a slice to return.